contrib/otelttrpc/interceptor.go

```diff
@@ -0,0 +1,197 @@
+/*
+   Copyright The containerd Authors.
+   Copyright The OpenTelemetry Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+
+   Based on go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc.
+*/
+
+package otelttrpc
+
+// ttrpc tracing middleware
+// https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/rpc.md
+import (
+	"context"
+	"net"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/baggage"
+	"go.opentelemetry.io/otel/codes"
+	semconv "go.opentelemetry.io/otel/semconv/v1.7.0"
+	"go.opentelemetry.io/otel/trace"
+	grpc_codes "google.golang.org/grpc/codes"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/containerd/ttrpc"
+	"github.com/containerd/ttrpc/contrib/otelttrpc/internal"
+)
+
+const (
+	// instrumentationName is the name of this instrumentation package.
+	instrumentationName = "github.com/containerd/ttrpc/contrib/otelttrpc"
+)
+
+type messageType attribute.KeyValue
+
+// Event adds an event of the messageType to the span associated with the
+// passed context with size (if message is a proto message).
+func (m messageType) Event(ctx context.Context, message interface{}) {
+	span := trace.SpanFromContext(ctx)
+	if p, ok := message.(proto.Message); ok {
+		span.AddEvent("message", trace.WithAttributes(
+			attribute.KeyValue(m),
+			semconv.MessageUncompressedSizeKey.Int(proto.Size(p)),
```

Comment:
This led to performance issue for otel grpc as the otlpreceiver consume lots of cpu for proto.Size. It has been removed since https://github.com/open-telemetry/opentelemetry-go-contrib/pull/3168